6_predict/src/train_models.R

```diff
@@ -16,11 +18,50 @@ split_data <- function(data, train_prop){
   training <- training(split)
   testing <- testing(split)
   
+  ##assign nested group ID to training and testing data sets
+  training <- merge(training, nested_groups, by.x= "GAGES_ID", by.y = "ID")
+  testing <- merge(testing, nested_groups, by.x = "GAGES_ID", by.y = "ID")
+  
+  ##find any nested group ids that are in both data sets
+  intersection_ids<-intersect(training$nested_group_id, testing$nested_group_id)
+  
+  ##find all un-nested group ids
+  unnested<- nested_groups %>%
+    group_by(nested_group_id) %>%
+    count()%>%
+    filter(n ==1)%>%
+    select(nested_group_id)
+  
+  ##identify nested gages to move from testing to training and
+  ##unnested replacements to move from training to testing
+  move_gages<- testing %>%
+    filter(nested_group_id %in% intersection_ids)
+  
+  replacement_gages<-training %>%
+    filter(nested_group_id %in% unnested$nested_group_id)%>%
+    sample_n(size = nrow(move_gages))
```

Comment:
So my initial thought on this was that it was better to have the nested basins in the training group because I feel like having potential duplicates in the testing set, which is smaller, might cause more bias than in the training group.  So I moved all the nested basins that got split up to the training data.  But, if we want to keep that initial split proportion of nested basins the same, maybe I should only replace them with other nested basins?

Reply:
By duplicates, do you mean nested gages?

That is a good point about trying to keep the test set free of nested gages. It could be challenging to force the proportion to be the same because it might be that it's not possible to do that. I'm okay with leaving it as-is for now. I think the larger problem (not for us) is how to consider in the loss function that nested gages are not independently contributing to the loss